src/features/unified-share-modal/UnifiedShareForm.js

```diff
@@ -54,15 +69,70 @@ class UnifiedShareForm extends React.Component<USFProps, State> {
         super(props);
 
         this.state = {
+            classificationLabelId: '',
             emailSharedLinkContacts: [],
             inviteCollabsContacts: props.initiallySelectedContacts,
             inviteePermissionLevel: '',
             isEmailLinkSectionExpanded: false,
+            isFetchingJustificationReasons: false,
             isInviteSectionExpanded: !!props.initiallySelectedContacts.length,
+            justificationReasons: [],
             showCollaboratorList: false,
         };
     }
 
+    componentDidUpdate(prevProps: USFProps) {
+        const { isCollabRestrictionJustificationAllowed, item, restrictedExternalCollabEmails } = this.props;
+        const {
+            restrictedExternalCollabEmails: prevRestrictedExternalCollabEmails,
+            isCollabRestrictionJustificationAllowed: prevIsCollabRestrictionJustificationAllowed,
+        } = prevProps;
+
+        const didExternalCollabRestrictionsChange =
+            !isEqual(restrictedExternalCollabEmails, prevRestrictedExternalCollabEmails) ||
+            isCollabRestrictionJustificationAllowed !== prevIsCollabRestrictionJustificationAllowed;
+
+        if (didExternalCollabRestrictionsChange && this.shouldRequireExternalCollabJustification()) {
+            this.fetchJustificationReasons(item, JUSTIFICATION_CHECKPOINT_EXTERNAL_COLLAB);
+        }
+    }
+
+    fetchJustificationReasons = (item: Item, checkpoint: JustificationCheckpoint) => {
```

Comment:
Using a pattern similar to what we have for `getContactsByEmail`, where the results are only relevant to the component and thus don't need to be saved in a store and passed down as props. Currently there is no need to cache or share justification reasons beyond this component, so I'm avoiding having them on a store on the webapp. Let me know in case there's a preference for that approach though.